kale/pipeline/domain_adapter.py

```diff
@@ -869,66 +801,81 @@ def critic_update_steps(self, batch):
+            critic_opt.zero_grad()
+
+            # update discriminator opt every k_critic steps
+            if (batch_id + 1) % self._k_critic == 0:
+                task_opt.step()
+                task_opt.zero_grad()
+
+        # log_metrics = get_aggregated_metrics_from_dict(log_metrics)
+        # log_metrics.update(get_metrics_from_parameter_dict(self.get_parameters_watch_list(), loss.device))
         log_metrics["train_total_loss"] = loss
         log_metrics["train_task_loss"] = task_loss
 
-        for key in log_metrics:
-            self.log(key, log_metrics[key])
+        self.log_dict(log_metrics, on_step=True, on_epoch=False)
 
-        return {
-            "loss": loss,  # required, for backward pass
-            # "progress_bar": {"class_loss": task_loss},
-            # "log": log_metrics,
-        }
+        return loss  # required, for backward pass
 
     # Add on_tpu=False etc following https://github.com/PyTorchLightning/pytorch-lightning/issues/2934
     # to fix error for WDGRLMod: TypeError: optimizer_step() got an unexpected keyword argument 'on_tpu'
-    def optimizer_step(
-        self,
-        current_epoch,
-        batch_nb,
-        optimizer,
-        optimizer_i,
-        second_order_closure=None,
-        on_tpu=False,
-        using_native_amp=False,
-        using_lbfgs=False,
-    ):
-        if current_epoch < self._init_epochs:
-            # do not update critic
-            if optimizer_i == 0:
-                pass
-            if optimizer_i == 1:
-                optimizer.step()
-                optimizer.zero_grad()
-        else:
-            if optimizer_i == 0:
-                optimizer.step()
-                optimizer.zero_grad()
-
-            # update discriminator opt every k_critic steps
-            if optimizer_i == 1:
-                if (batch_nb + 1) % self._k_critic == 0:
-                    optimizer.step()
-                optimizer.zero_grad()
-
-        optimizer.step(closure=second_order_closure)
+    # def optimizer_step(
+    #     self,
+    #     current_epoch,
+    #     batch_nb,
+    #     optimizer,
+    #     # optimizer_i,
+    #     second_order_closure=None,
+    #     # on_tpu=False,
+    #     # using_native_amp=False,
+    #     # using_lbfgs=False,
+    # ):
+    #     if current_epoch < self._init_epochs:
+    #         # do not update critic
+    #         if optimizer_i == 0:
+    #             pass
+    #         if optimizer_i == 1:
+    #             optimizer.step()
+    #             optimizer.zero_grad()
+    #     else:
+    #         if optimizer_i == 0:
+    #             optimizer.step()
+    #             optimizer.zero_grad()
+    #
+    #         # update discriminator opt every k_critic steps
+    #         if optimizer_i == 1:
+    #             if (batch_nb + 1) % self._k_critic == 0:
+    #                 optimizer.step()
+    #             optimizer.zero_grad()
+    #
+    #     optimizer.step(closure=second_order_closure)
```

Comment:
Similar Q: keep vs remove

Reply:
Keep but add a new comment at the top for description.

`# PyTorch Lightning 2.0+ has a different optimizer_step, so we implement this staged optimization in training_step above. We will retain optimizer_step until we have assessed WDGRLTrainerMod, after which it will be removed.`